### ISSUE TITLE
Defaults for optional maps and lists containing simple types

### ISSUE THREAD

**Issue by codello** [hashicorp/terraform#27622]
### Terraform Version
```
Terraform v0.14.5
```

### Terraform Configuration Files
```terraform
terraform {
  experiments = [module_variable_optional_attrs]
}

variable "test" {
  type = object({
    m = optional(map(string))
    l = optional(list(string))
  })
  default = {}
}

locals {
  d = defaults(var.test, {
    m = {}
    l = []
  })
}

output "o" {
  value = local.d
}
```

### Debug Output
[Gist](https://gist.github.com/codello/6f8266358bce501bb296e5b3d411efbd), see `terraform-empty.log`.

### Expected Behavior
```
Changes to Outputs:
  + o = {
      + m = {}
      + l = []
    }
```

### Actual Behavior
See `Debug Output`. Changing the `defaults` values for `m` or `l` to either `null` or any string value causes the following error
```
Call to function "defaults" failed: panic in function implementation: can't
use ElementIterator on null value
```
The full debug output for the `null` case is included in the above gist as `terraform-null.log`.

### Steps to Reproduce
1. Copy and paste code from `Terraform Configuration Files`
2. `terraform init`
3. `terraform plan`

**Comment by apparentlymart** [comment]
Hi @codello! Thanks for reporting this.

It looks like you've encountered the same problem as reported in #27385, so I'm going to close this one just to consolidate the discussion over there. Thanks again!

---

While it's certainly incorrect for this to panic rather than returning a better error, note that the `defaults` handling for lists and maps is for the default to be a default value _for each element_, rather than a default value for the list or map as a whole, so the default values for `m` and `l` should both be strings to match the element type of each collection.

Using a default in this way for a list or set with a primitive element type is not particularly useful, but this behavior is there to support setting default values for nested attributes when the input is a list or map of an _object_ type. The handling of primitive types is derived from the behavior for object types, and so the following (not very useful in practice) behavior emerges from those mapping rules:

If you set the default value for `m` to `"hello"` and then set `m` to `[null, "world"]`, the `defaults` function will see that the first element is `null` and replace it with the default value, producing `["hello", "world"]`.

I think the main thing to consider for this, which I intend to investigate further in #27385, is how `defaults` ought to handle a collection-type attribute that isn't specified at all. In this case it looks like you wanted `defaults` to replace the `null` with an empty collection, which I think would be a reasonable thing for `defaults` to do automatically, because a `null` list or map is not something we typically expect to deal with in Terraform expressions. I'll investigate that angle more when I have some more time to work on #27385.

**Comment by codello** [comment]
Thank you for the quick and extensive response. I must have missed that issue when looking for similar issues, my bad. Thanks as well for the explanation of the `defaults` function. I actually stumbled across this issue when I was looking for a way to default an `optional(map)` to an empty map. I agree that `defaults` replacing `null` with an empty map or list respectively would make sense (although there might be times where a `null` map or list might come in handy).

**Comment by korinne** [comment]
Hi all, we recently released the [Terraform v1.3 alpha](https://releases.hashicorp.com/terraform/1.3.0-alpha20220608), which includes the ability to mark object type attributes as optional and set default values.

I'm following up with issues that have provided feedback on the previous experiment, and wanted to invite you all to try the alpha and provide any feedback on this updated design. You can learn more/add comments[ here](https://discuss.hashicorp.com/t/request-for-feedback-optional-object-type-attributes-with-defaults-in-v1-3-alpha/40550). Thank you so much in advance, and hope you like the feature!
